sampling/chain_randomness.go

```diff
@@ -6,62 +6,57 @@ import (
 	"github.com/filecoin-project/go-filecoin/types"
 )
 
-// LookbackParameter is the protocol parameter defining how many blocks in the
-// past to look back to sample randomness values.
+// LookbackParameter defines how many non-empty tiptsets (not rounds) earlier than any sample
+// height (in rounds) from which to sample the chain for randomness.
+// This constant is a protocol (actor) parameter and should be defined in actor code.
 const LookbackParameter = 3
 
-// SampleChainRandomness produces a slice of random bytes sampled from a tip set
-// in the provided slice of tip sets at a given height (minus lookback). This
-// function assumes that the tip set slice is sorted by block height in
-// descending order.
-//
-// SampleChainRandomness is useful for things like PoSt challenge seed
-// generation.
-func SampleChainRandomness(sampleHeight *types.BlockHeight, tipSetsSortedByBlockHeightDescending []types.TipSet) ([]byte, error) {
+// SampleChainRandomness produces a slice of bytes (a ticket) sampled from the tipset `LookbackParameter`
+// tipsets (not rounds) prior to the highest tipset with height less than or equal to `sampleHeight`.
+// The tipset slice must be sorted by descending block height.
+func SampleChainRandomness(sampleHeight *types.BlockHeight, tipSetsDescending []types.TipSet) ([]byte, error) {
+	// Find the first (highest) tipset with height less than or equal to sampleHeight.
+	// This is more complex than necessary: https://github.com/filecoin-project/go-filecoin/issues/3025
 	sampleIndex := -1
-	tipSetsLen := len(tipSetsSortedByBlockHeightDescending)
-	lastIdxInTipSets := tipSetsLen - 1
-
-	for i := 0; i < tipSetsLen; i++ {
-		height, err := tipSetsSortedByBlockHeightDescending[i].Height()
+	for i, tip := range tipSetsDescending {
+		height, err := tip.Height()
 		if err != nil {
 			return nil, errors.Wrap(err, "error obtaining tip set height")
 		}
 
-		if types.NewBlockHeight(height).Equal(sampleHeight) {
+		if types.NewBlockHeight(height).LessEqual(sampleHeight) {
```

Comment:
I don't think we should do this.  While it will probably work for some time this is not compatible with a finished PoSt generate/verify system.  We need to sample an exact seed from an exact block height so that proving and verifying interoperate correctly and we should be able to provide it every time.  I believe the root of this bug is that the miner actor is getting into a state where its `provingPeriodStart` is not 100% for sure associated with a tipset blockheight that exists.  We should address this directly at the root.

I'm looking into this.  cc @acruikshank I can't see any issue with the change [here](https://github.com/filecoin-project/go-filecoin/commit/331e559db35cc2c6c98b419c0464cc1ee2d2d73e) but it is my primary suspect.

Reply:
Looking closer at the commit mentioned above I see that we *do* have a consistent reference seed.  In light of that this change is ok, especially since all seed sampling code is going to be seeing some major improvements in the near future.

Reply:
I agree with the second comment. This will generate a consistent seed and is the correct approach.